### PR TITLE
fix(components): [select] Filter duplicate emits of "update options"

### LINKED ISSUE #11890: [Component] [select] 2.2.35最新版的ElSelect中使用slot会造成内存溢出死循环
4280C503-F6EA-4405-AA70-6C4811E711C1
4280C503-F6EA-4405-AA70-6C4811E711C2

Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.47`
- Element Plus Version: `2.2.35`
- Browser / OS: `Chrome 110.0.5481.177（正式版本） (x86_64)`
- Build Tool: `Vue CLI`

### Related Component

- `el-select`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtc2VsZWN0IHYtbW9kZWw9XCJ2YWx1ZVwiIHBsYWNlaG9sZGVyPVwiU2VsZWN0XCI+XG4gICAgPHRlbXBsYXRlIHYtZm9yPVwiaXRlbSBpbiBjaXRpZXNcIiA6a2V5PVwiaXRlbS52YWx1ZVwiPlxuICAgIDxlbC1vcHRpb25cbiAgICAgIDpsYWJlbD1cIml0ZW0ubGFiZWxcIlxuICAgICAgOnZhbHVlPVwiaXRlbS52YWx1ZVwiXG4gICAgPlxuICAgICAgPHNsb3QgOml0ZW09XCJpdGVtXCI+XG5cdCAgICAgIDxzcGFuIHN0eWxlPVwiZmxvYXQ6IGxlZnRcIj57eyBpdGVtLmxhYmVsIH19PC9zcGFuPlxuICAgICAgICA8c3BhblxuICAgICAgICAgIHN0eWxlPVwiXG4gICAgICAgICAgICBmbG9hdDogcmlnaHQ7XG4gICAgICAgICAgICBjb2xvcjogdmFyKC0tZWwtdGV4dC1jb2xvci1zZWNvbmRhcnkpO1xuICAgICAgICAgICAgZm9udC1zaXplOiAxM3B4O1xuICAgICAgICAgIFwiXG4gICAgICAgICAgPnt7IGl0ZW0udmFsdWUgfX08L3NwYW5cbiAgICAgICAgPlxuICAgICAgPC9zbG90PlxuICAgIDwvZWwtb3B0aW9uPlxuICAgIDwvdGVtcGxhdGU+XG4gIDwvZWwtc2VsZWN0PlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdCBsYW5nPVwidHNcIiBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgdmFsdWUgPSByZWYoJycpXG5jb25zdCBjaXRpZXMgPSBbXG4gIHtcbiAgICB2YWx1ZTogJ0JlaWppbmcnLFxuICAgIGxhYmVsOiAnQmVpamluZycsXG4gIH0sXG4gIHtcbiAgICB2YWx1ZTogJ1NoYW5naGFpJyxcbiAgICBsYWJlbDogJ1NoYW5naGFpJyxcbiAgfSxcbiAge1xuICAgIHZhbHVlOiAnTmFuamluZycsXG4gICAgbGFiZWw6ICdOYW5qaW5nJyxcbiAgfSxcbiAge1xuICAgIHZhbHVlOiAnQ2hlbmdkdScsXG4gICAgbGFiZWw6ICdDaGVuZ2R1JyxcbiAgfSxcbiAge1xuICAgIHZhbHVlOiAnU2hlbnpoZW4nLFxuICAgIGxhYmVsOiAnU2hlbnpoZW4nLFxuICB9LFxuICB7XG4gICAgdmFsdWU6ICdHdWFuZ3pob3UnLFxuICAgIGxhYmVsOiAnR3Vhbmd6aG91JyxcbiAgfSxcbl1cbjwvc2NyaXB0PlxuIiwiaW1wb3J0X21hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJfbyI6e319)

### Steps to reproduce

最新版本2.2.35里select中如果使用slot会造成内存溢出页面直接卡崩，官方demo也是一样，过往版本使用正常，只有最新版本2.2.35会这样

```vue
<template>
 <el-select v-model="value" placeholder="Select">
    <template v-for="item in cities" :key="item.value">
    <el-option
      :label="item.label"
      :value="item.value"
    >
      <slot :item="item">
	<span style="float: left">{{ item.label }}</span>
        <span>{{ item.value }}</span
        >
      </slot>
    </el-option>
    </template>
  </el-select>
</template>

### What is Expected?

正常使用

### What is actually happening?

内存溢出，死循环页面直接卡崩

### Additional comments

_(empty)_

<!-- 如果你需要插入截图，请在这粘贴。 -->

### PR DESCRIPTION
**Pull request by godxiaoji**
closed #11890 

Another problem is found in https://github.com/element-plus/element-plus/issues/11878.

For example:
```vue
<script lang="tsx">
import { defineComponent, ref } from 'vue'
export default defineComponent({
  setup() {
    const options = ref([
      {
        value: 'Option1',
        label: 'Option1',
      },
      {
        value: 'Option2',
        label: 'Option2',
      },
    ])

    return () => (
      <el-select>
        {options.value.map((item) => (
          <el-option key={item.value} label={item.label} value={item.value} />
        ))}
      </el-select>
    )
  },
})
</script>

```

The following problems occur when using jsx:

image

Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by godxiaoji** [comment]
This PR is based on #11868 changes

**Review by jw-foss — APPROVED** [review]

File: packages/components/select/src/options.ts [hunk 1]
```diff
@@ -6,6 +6,21 @@ export default defineComponent({
   name: 'ElOptions',
   emits: ['update-options'],
   setup(_, { slots, emit }) {
+    let oldOptions: any[] = []
```
Comment (jw-foss) [inline packages/components/select/src/options.ts:9]:
```suggestion
    let cachedOptions: any[] = []
```

File: packages/components/select/src/options.ts [hunk 2]
```diff
@@ -6,6 +6,21 @@ export default defineComponent({
...
+      if (a.length === b.length) {
+        for (const [index] of a.entries()) {
+          if (a[index] != b[index]) {
+            return false
+          }
+        }
+      } else {
+        return false
+      }
+      return true
```
Comment (jw-foss) [inline packages/components/select/src/options.ts:21]:
```suggestion
      if (a.length !== b.length) return false
      for (const [index] of a.entries()) {
          if (a[index] != b[index]) {
            return false
          }
      }
      return true
```

**Review by jw-foss — COMMENTED** [review]
@godxiaoji thank you for solving this, I left a few comments for you, let me know if you have any questions 👍

**Review by jw-foss — APPROVED** [review]

**Comment by godxiaoji** [comment]
> @godxiaoji thank you for solving this, I left a few comments for you, let me know if you have any questions 👍

The code has been standardized according to your suggestion.

**Comment by jw-foss** [comment]
Merged 🚀

**Pull request merged**